### PR TITLE
Add C++ generation support to gr-digital

### PR DESCRIPTION
**Pull request by aru31**
PR for issue #2297 to add add cpp support to gr-digital

### CODE REVIEW

File: gr-digital/grc/digital_ofdm_rx.block.yml [hunk 1]
```diff
@@ -108,4 +109,37 @@ templates:
             debug_log=${log},
             scramble_bits=${scramble_bits})
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/ofdm_equailizer_base.h>',
```
Comment (haakov) [inline gr-digital/grc/digital_ofdm_rx.block.yml]: Typo: misspelled "equalizer"

File: gr-digital/grc/digital_ofdm_rx.block.yml [hunk 2]
```diff
@@ -108,4 +109,37 @@ templates:
...
+            % if sync_word1:
+            sync_word1=${sync_word1},
+            % endif
+            % if sync_word2:
+            sync_word2=${sync_word2},
+            % endif
+            bps_header=${header_mod.bps},
+            bps_payload=${payload_mod.bps},
+            debug_log=${log},
+            scramble_bits=${scramble_bits})
```
Comment (haakov) [inline gr-digital/grc/digital_ofdm_rx.block.yml]: Typo: missing semicolons

File: gr-digital/grc/digital_ofdm_tx.block.yml [hunk 3]
```diff
@@ -114,4 +115,37 @@ templates:
             debug_log=${log},
             scramble_bits=${scramble_bits})
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/ofdm_equailizer_base.h>',
```
Comment (haakov) [inline gr-digital/grc/digital_ofdm_tx.block.yml]: Typo: misspelled "equalizer"

File: gr-digital/grc/digital_protocol_parser_b.block.yml [hunk 4]
```diff
@@ -19,4 +20,10 @@ templates:
     imports: from gnuradio import digital
     make: digital.protocol_parser_b(${format})
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/protocol_parser_b.h>']
+    declarations: 'digital::protocol_parser_b::sptr ${id};'
+make: 'this->${id} = fft::ctrlport_probe_psd::make(${format});'
```
Comment (haakov) [inline gr-digital/grc/digital_protocol_parser_b.block.yml]: Typo: missing indent

File: gr-digital/grc/variable_header_format_default.block.yml [hunk 5]
```diff
@@ -27,4 +28,17 @@ templates:
         ${threshold}, ${bps})
         % endif
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/header_format_default.h>']
+    declarations: 'digital::header_format_default::sptr ${id};'
+    var_make: |-
+        % if int(access_code)==0:
+        self.${id} = ${id} = digital.header_format_default(digital.packet_utils.default_access_code,\
+        ${threshold}, ${bps})
```
Comment (haakov) [inline gr-digital/grc/variable_header_format_default.block.yml]: Typo: missing semicolons

File: gr-digital/grc/digital_ofdm_rx.block.yml (hunk 1 above)
Comment (aru31) [reply gr-digital/grc/digital_ofdm_rx.block.yml]: Done

File: gr-digital/grc/digital_ofdm_rx.block.yml (hunk 2 above)
Comment (aru31) [reply gr-digital/grc/digital_ofdm_rx.block.yml]: Done

File: gr-digital/grc/digital_ofdm_tx.block.yml (hunk 3 above)
Comment (aru31) [reply gr-digital/grc/digital_ofdm_tx.block.yml]: Done

File: gr-digital/grc/digital_protocol_parser_b.block.yml (hunk 4 above)
Comment (aru31) [reply gr-digital/grc/digital_protocol_parser_b.block.yml]: Done

File: gr-digital/grc/digital_constellation_rect.block.yml [hunk 6]
```diff
@@ -55,4 +56,18 @@ templates:
         self.${id}.set_soft_dec_lut(${soft_dec_lut}, ${precision})
         % endif
 
+cpp_templates:
+    includes: []
+    declarations: 'digital::constellation_rect::sptr ${id};'
+    var_make: |-
+        self.${id} = ${id} = digital.constellation_rect(${const_points}, ${sym_map},
+        ${rot_sym}, ${real_sect}, ${imag_sect}, ${w_real_sect}, ${w_imag_sect}).base()
```
[the remark this answers is not in the document]
Comment (aru31) [reply gr-digital/grc/digital_constellation_rect.block.yml]: Done

File: gr-digital/grc/digital_modulate_vector.block.yml [hunk 7]
```diff
@@ -19,4 +20,11 @@ templates:
     var_make: self.${id} = ${id} = digital.modulate_vector_bc(${mod} .to_basic_block(),
         ${data}, ${taps})
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/modulate_vector.h>']
+    declarations: 'digital::modulate_vector_bc::sptr ${id};'
+    var_make: self.${id} = ${id} = digital.modulate_vector_bc(${mod} .to_basic_block(),
```
[the remark this answers is not in the document]
Comment (aru31) [reply gr-digital/grc/digital_modulate_vector.block.yml]: Done

File: gr-digital/grc/variable_header_format_default.block.yml (hunk 5 above)
Comment (aru31) [reply gr-digital/grc/variable_header_format_default.block.yml]: Done

File: gr-digital/grc/digital_constellation.block.yml [hunk 8]
```diff
@@ -55,4 +56,21 @@ templates:
         self.${id}.set_soft_dec_lut(${soft_dec_lut}, ${precision})
         % endif
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/constellation.h>']
+    declarations: 'digital::constellation::sptr ${id};'
+    var_make: |-
+        % if str(type) == "calcdist":
+        self.${id} = ${id} = digital.constellation_calcdist(${const_points}, ${sym_map},
+        ${rot_sym}, ${dims}).base()
```
[the remark this answers is not in the document]
Comment (aru31) [reply gr-digital/grc/digital_constellation.block.yml]: Done

File: gr-digital/grc/digital_constellation_modulator.block.yml [hunk 9]
```diff
@@ -1,5 +1,6 @@
 id: digital_constellation_modulator
 label: Constellation Modulator
+flags: [ python, cpp ]
```
Comment (haakov) [inline gr-digital/grc/digital_constellation_modulator.block.yml]: The cpp flag should also be removed

File: gr-digital/grc/digital_constellation_modulator.block.yml [hunk 10]
```diff
@@ -54,4 +55,18 @@ templates:
             verbose=${verbose},
             log=${log})
 
+cpp_templates:
+    includes: []
```
[the remark this answers is not in the document]
Comment (aru31) [reply gr-digital/grc/digital_constellation_modulator.block.yml]: Done

File: gr-digital/grc/digital_constellation_modulator.block.yml (hunk 9 above)
Comment (aru31) [reply gr-digital/grc/digital_constellation_modulator.block.yml]: Done

File: gr-digital/grc/digital_gfsk_demod.block.yml [hunk 11]
```diff
@@ -1,6 +1,6 @@
 id: digital_gfsk_demod
 label: GFSK Demod
-flags: [ python, cpp ]
+flags: [ pythonn]
```
Comment (haakov) [inline gr-digital/grc/digital_gfsk_demod.block.yml]: Typo

**Comment by aru31** [comment]
Shall I merge all these in one single commit?

**Comment by haakov** [comment]
Okay, sweet! So what remains now is the tedious part. 

There are a lot of blocks with simple parameters, these typically don't need more work. By simple, I mean parameters that have similar representations in Python and C++. These are mainly `real` and `int` dtypes. The `hex` dtype might work right away as well, but I haven't tested it. Most blocks also have an enum with the id "type", this is usually fine left as it is.

However, some parameters have dtypes such as:

- Enums with `True` and `False` values (Needs to be uncapitalized for C++)
- Parameters such as the Signal Source's Waveform parameter, with values such as `analog.GR_CONST_WAVE` (Change `.` to `::` and it's okay)
- All kinds of vectors

These need some kind of "conversion" to work in a C++ file, and the `translations` key is created for this purpose. I suggest you try to generate some C++ flowgraphs with these blocks, try to build and run them. After fixing a couple of files, you'll see what works and what doesn't. You can look at how I've used `translations` in other blocks. Oh, and please ask if you have any questions :)

PS: You'll want to apply this fix before attempting to generate any C++ flowgraphs: https://github.com/gnuradio/gnuradio/pull/2336

File: gr-digital/grc/digital_ofdm_sync_sc_cfb.block.yml [hunk 12]
```diff
@@ -42,4 +43,20 @@ templates:
...
+    declarations: 'digital::ofdm_sync_sc_cfb::sptr ${id};'
+    make: |-
+        this->${id} = digital::ofdm_sync_sc_cfb::make(
+          ${fft_len},
+          ${cp_len},
+          ${use_even_carriers},
+          ${threshold});
+    link: ['gnuradio-digital']
+    callbacks:
+      - set_treshold(${threshold})
```
Comment (haakov) [inline gr-digital/grc/digital_ofdm_sync_sc_cfb.block.yml]: Indentation typo, causes error on GRC startup

File: gr-digital/grc/digital_cpmmod_bc.block.yml [hunk 13]
```diff
@@ -37,4 +38,20 @@ templates:
...
+    make: |-
+        this->${id} = digital::cpmmod_bc::make(
+            ${type},
+            ${mod_index},
+            ${samples_per_symbol},
+            ${L},
+            ${beta});
+    link: ['gnuradio-digital']
+    translations:
+        analog.cpm: 'cpm('
```
Comment (haakov) [inline gr-digital/grc/digital_cpmmod_bc.block.yml]: I'm afraid this won't work. You can find the parameters in the header file: https://github.com/gnuradio/gnuradio/blob/7074f054c25a69392e7ae4e1f0204a1c5ce242f8/gr-digital/include/gnuradio/digital/cpmmod_bc.h#L73

File: gr-digital/grc/digital_ofdm_sync_sc_cfb.block.yml (hunk 12 above)
Comment (aru31) [reply gr-digital/grc/digital_ofdm_sync_sc_cfb.block.yml]: Actually, I checked out the master branch of gnuradio and found that this particular indentation and a typo was already there. So, Fixed that :)

File: gr-digital/grc/digital_burst_shaper.block.yml [hunk 14]
```diff
@@ -47,4 +48,19 @@ templates:
     make: digital.burst_shaper_${type.fcn}(${window}, ${pre_padding}, ${post_padding},
         ${insert_phasing}, ${length_tag_name})
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/burst_shaper.h>']
+    declarations: 'digital::burst_shaper_${type.fcn}::sptr ${id};'
+    make: |-
+        this->${id} = digital::burst_shaper_${type.fcn}::make(
+                ${window},
```
Comment (haakov) [inline gr-digital/grc/digital_burst_shaper.block.yml]: `window` is a vector

File: gr-digital/grc/digital_chunks_to_symbols.block.yml [hunk 15]
```diff
@@ -56,4 +57,15 @@ templates:
     callbacks:
     - set_symbol_table(${symbol_table})
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/chunks_to_symbols.h>']
+    declarations: 'digital::chunks_to_symbols_${in_type.fcn}${out_type.fcn}::sptr ${id};'
+    make: |-
+        this->${id} = digital::chunks_to_symbols_${in_type.fcn}${out_type.fcn}::make(
+                ${symbol_table},
```
Comment (haakov) [inline gr-digital/grc/digital_chunks_to_symbols.block.yml]: `symbol_table` is a vector

File: gr-digital/grc/digital_constellation_receiver_cb.block.yml [hunk 16]
```diff
@@ -59,4 +60,18 @@ templates:
...
+    includes: ['#include <gnuradio/digital/constellation_receiver_cb.h>']
+    declarations: 'digital::constellation_receiver_cb::sptr ${id};'
+    make: |-
+        this->${id} = digital::constellation_receiver_cb::make(
+            ${constellation},
+            ${loop_bw},
+            ${fmin},
+            ${fmax}));
+    link: ['gnuradio-digital']
+    translations:
```
Comment (haakov) [inline gr-digital/grc/digital_constellation_receiver_cb.block.yml]: Actually, this block doesn't need `translations `at all. `showports` is not mentioned in the templates.

File: gr-digital/grc/digital_constellation.block.yml [hunk 17]
```diff
@@ -55,4 +56,23 @@ templates:
...
         % endif
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/constellation.h>']
+    declarations: 'digital::constellation::sptr ${id};'
+    var_make: |-
+        std::vector<int> sym_map = {${str(sym_map)[1:-1]}};
+        std::vector<gr_complex> const_points = {${str(const_points)[1:-1]}};
+        % if str(type) == "calcdist":
+        self.${id} = ${id} = digital.constellation_calcdist(${const_points}, ${sym_map},
```
Comment (haakov) [inline gr-digital/grc/digital_constellation.block.yml]: The `${}` must be removed from `const_points `and `sym_map`

File: gr-digital/grc/digital_corr_est_cc.block.yml [hunk 18]
```diff
@@ -40,4 +41,22 @@ templates:
     - set_mark_delay(${mark_delay})
     - set_threshold(${threshold})
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/corr_est_cc.h>']
+    declarations: 'digital::corr_est_cc::sptr ${id};'
+    make: |-
+        std::vector<gr_complex> symbols = {${str(symbols)[1:-1]}};
+        this->${id} = digital::corr_est_cc::make(
+            ${symbols},
```
Comment (haakov) [inline gr-digital/grc/digital_corr_est_cc.block.yml]: `${}` must be removed from `symbols`

File: gr-digital/grc/variable_header_format_default.block.yml [hunk 19]
```diff
@@ -27,4 +28,17 @@ templates:
         ${threshold}, ${bps})
         % endif
 
+cpp_templates:
+    includes: ['#include <gnuradio/digital/header_format_default.h>']
+    declarations: 'digital::header_format_default::sptr ${id};'
+    var_make: |-
+        % if int(access_code)==0:
+        self.${id} = ${id} = digital.header_format_default(digital.packet_utils.default_access_code,\
```
Comment (haakov) [inline gr-digital/grc/variable_header_format_default.block.yml]: Looks like you forgot to convert this one :)

**Comment by marcusmueller** [comment]
hey, I lost track of this - is this ready for merge, @haakov ?

**Comment by haakov** [comment]
Sorry about the delay.

I don't think I noticed any other problems last time I read through it, I can take another look tomorrow.

**Comment by haakov** [comment]
Okay! One final thing, the OFDM TX and RX blocks appear to be Python-only, so the C++ templates for these should be removed. Otherwise, looks good :)

**Comment by aru31** [comment]
@haakov: changes made. Thanks a lot for your constant reviews. :)

**Comment by noc0lour** [comment]
This is good to go for now. Fixes can be applied later.

**Pull request merged**